### PR TITLE
Tuning documentation for foreman installer

### PR DESCRIPTION
**Pull request by sthirugn**

### CODE REVIEW

File: _includes/manuals/1.24/3.2.4_tuning_options.md
Comment (ehelms) [inline _includes/manuals/1.24/3.2.4_tuning_options.md]: Capitalize Foreman in the last sentence. Change customer to user perhaps?

File: _includes/manuals/1.24/3.2.4_tuning_options.md
Comment (ehelms) [inline _includes/manuals/1.24/3.2.4_tuning_options.md]: Might reword to `with a predefined tuning profile named default` to make it clearer.

Comment (ehelms) [reply _includes/manuals/1.24/3.2.4_tuning_options.md]: Consider changing `this` to `the default profile` to make it clearer

File: _includes/manuals/1.24/3.2.4_tuning_options.md
Comment (ekohl) [inline _includes/manuals/1.24/3.2.4_tuning_options.md]: I'd prefer to recommend `foreman-installer --help | grep tuning`.

File: _includes/manuals/1.24/3.2.4_tuning_options.md
Comment (ekohl) [inline _includes/manuals/1.24/3.2.4_tuning_options.md]: I think we can leave this out. This was part of the initial design but was never communicated. Adding it here will only confuse people. (There's also a typo in `/etc`)

**Review by ekohl — COMMENTED** [review]
General note: when we document things in Foreman, they should be included in `nightly`. When we branch the next version's documentation we use that as a base. If you only document it in `1.24`, it'll be lost in `2.0`.

Since the `--tuning` parameter is only available in Katello scenarios I think the main Foreman documentation would be misleading. Shall we move it to Katello's documentation?

File: _includes/manuals/1.24/3.2.4_tuning_options.md
Comment (ehelms) [reply _includes/manuals/1.24/3.2.4_tuning_options.md]: Does that show the current value?

Comment (ekohl) [reply _includes/manuals/1.24/3.2.4_tuning_options.md]: Note the `default:` part.

```console
# foreman-installer --help | grep tuning
    --tuning INSTALLATION_SIZE    Tune for an installation size. Choices: default, medium, large, extra-large, extra-extra-large (default: "default")
```
After changing it to large:
```console
# foreman-installer --help | grep tuning
    --tuning INSTALLATION_SIZE    Tune for an installation size. Choices: default, medium, large, extra-large, extra-extra-large (default: "large")
```

**Comment by sthirugn** [comment]
> General note: when we document things in Foreman, they should be included in `nightly`. When we branch the next version's documentation we use that as a base. If you only document it in `1.24`, it'll be lost in `2.0`.
> 
> Since the `--tuning` parameter is only available in Katello scenarios I think the main Foreman documentation would be misleading. Shall we move it to Katello's documentation?

I am not aware of the scenario differences in the installer. Let me know which repo to move this to, I can happily move it to.

File: _includes/manuals/1.24/3.2.4_tuning_options.md
Comment (sthirugn) [reply _includes/manuals/1.24/3.2.4_tuning_options.md]: oh wait, this totally beats the `default` purpose.  I thought `default` is something which is really `default`, do we really  use it to show the current configured value across the installer?

File: _includes/manuals/1.24/3.2.4_tuning_options.md
Comment (sthirugn) [reply _includes/manuals/1.24/3.2.4_tuning_options.md]: Many users are used to seeing override in custom-hiera.yml for tuning profile. Hence I added it. This was also brought out by [QE](https://bugzilla.redhat.com/show_bug.cgi?id=1772390#c0). In my new commit, I added a new that the users can still use custom-hiera.yml to override any configuration. Let me know what you think.

**Comment by ekohl** [comment]
> I am not aware of the scenario differences in the installer. Let me know which repo to move this to, I can happily move it to.

Katello is documented at https://github.com/theforeman/theforeman.org/tree/gh-pages/plugins/katello.

File: _includes/manuals/1.24/3.2.4_tuning_options.md
Comment (ekohl) [reply _includes/manuals/1.24/3.2.4_tuning_options.md]: The word "default" is a bit overloaded. In this case `default:` means the default value for the command line parameter if not specified. So `default: "default"` means the default tuning is configured.

File: _includes/manuals/1.24/3.2.4_tuning_options.md
Comment (ekohl) [reply _includes/manuals/1.24/3.2.4_tuning_options.md]: Good point. It may also be good to recommend users who have previously tuned manually should review their values. They can overwrite built in tuning but is not recommended.

File: _includes/manuals/1.24/3.2.4_tuning_options.md
Comment (sthirugn) [reply _includes/manuals/1.24/3.2.4_tuning_options.md]: Updated now

**Comment by sthirugn** [comment]
I addressed all the review comments except the one for moving this doc to the katello repository.

**Comment by sthirugn** [comment]
> > I am not aware of the scenario differences in the installer. Let me know which repo to move this to, I can happily move it to.
> 
> Katello is documented at https://github.com/theforeman/theforeman.org/tree/gh-pages/plugins/katello.

Ewoud, foreman users would need to do tuning as well.  Should we just keep the doc here in foreman but add a note that some configuration may be purely katello related?

**Comment by ekohl** [comment]
> Ewoud, foreman users would need to do tuning as well. Should we just keep the doc here in foreman but add a note that some configuration may be purely katello related?

The whole option `--tuning` doesn't exist when using the `foreman` scenario so it's not really possible to use it. On a technical level is can easily be implemented as well but due to time constraints this wasn't done.

**Comment by ehelms** [comment]
Sounds like for 1.24, move the docs to Katello only since thats the only scenario this works for. But we should file an issue (https://projects.theforeman.org/projects/puppet-foreman/issues/new) to enable this for Foreman scenario for 2.0 or 2.1.

**Comment by sthirugn** [comment]
2 things updated:
* Moved the tuning options doc to katello 3.14 now
* created a new [`issue`](https://projects.theforeman.org/issues/28352) to track the doc update for Foreman 2.0/2.1

File: plugins/katello/3.14/installation/index.md
Comment (ekohl) [inline plugins/katello/3.14/installation/index.md:206]: On Github you can use `console` as well and use `#` (or `$`, but this should run as root) to separate the command from output:
```console
# foreman-installer --help | grep tuning
     --tuning INSTALLATION_SIZE    Tune for an installation size. Choices: default, medium, large, extra-large, extra-extra-large (default: "default")
```

Don't know if our website renders it in the same way.

File: plugins/katello/3.14/installation/index.md
Comment (ekohl) [inline plugins/katello/3.14/installation/index.md]: I'm wondering if it makes sense to specify all options since it's essentially the same.

File: plugins/katello/3.14/installation/index.md
Comment (ekohl) [inline plugins/katello/3.14/installation/index.md]: https://github.com/theforeman/foreman-installer/commit/8f8f2beaf46ed1658a28b3877a571054a5c5cec6 makes these a hard lower limits and need `--disable-system-checks` to skip. Maybe worth mentioning.

**Review by ekohl — COMMENTED** [review]
Katello has the same thing with a nightly and 3.14. I'm fine with first applying it in one place and only copying when you're done to avoid duplicating the work but something that should be fixed before merging.

File: plugins/katello/3.14/installation/index.md
Comment (sthirugn) [reply plugins/katello/3.14/installation/index.md]: I wanted to make it super clear, but it feels redundant - I can combine them - providing just one example.

File: plugins/katello/3.14/installation/index.md
Comment (sthirugn) [reply plugins/katello/3.14/installation/index.md:206]: yes I have seen it in other projects.  But I don't see any reference of it being used in this project. So I used the `bash` format.

File: plugins/katello/3.14/installation/index.md
Comment (sthirugn) [reply plugins/katello/3.14/installation/index.md]: good point, will add

**Comment by sthirugn** [comment]
> Katello has the same thing with a nightly and 3.14. I'm fine with first applying it in one place and only copying when you're done to avoid duplicating the work but something that should be fixed before merging.

Good to know. Updated nightly now.

**Review by ehelms — APPROVED** [review]
@ekohl final review is yours

File: plugins/katello/3.14/installation/index.md
Comment (ekohl) [inline plugins/katello/3.14/installation/index.md]: This may be a bit misleading. As it's written now it implies when `medium` is selected, `common` is no longer applied which isn't true. Common is used for all tuning. In a mathematical way: `tuning = common + profile`. Technically Hiera looks at `foreman.hiera/tuning/sizes/default.yaml` and then `foreman.hiera/tuning/common.yaml`. When a file doesn't exist, Hiera just continues.

Comment (ehelms) [reply plugins/katello/3.14/installation/index.md]: This is true for now unless we make an empty default yaml file. So for me it reads OK.

Comment (ekohl) [reply plugins/katello/3.14/installation/index.md]: I disagree. `common` is also used with `medium`. The current text implies you **only** need to look at it when you use default. Given later on we recommend users to look at their tuning, they may keep redundant tuning in `custom-hiera.yaml`.

I'd split it up and leave out how to find the exact tuning till later.

```suggestion
When the foreman-installer is run, it is deployed with a predefined tuning profile named `default`. Other than the default tuned profile, foreman-installer supports 4 different tuning profiles:
```

File: plugins/katello/3.14/installation/index.md
Comment (ekohl) [inline plugins/katello/3.14/installation/index.md]:
```suggestion
Based on your environment needs, use one of the tuning profiles (`medium`, `large`, `extra-large`, `extra-extra-large`) in the installer.  For example, `medium` profile can be applied like:
```

File: plugins/katello/3.14/installation/index.md
Comment (ekohl) [inline plugins/katello/3.14/installation/index.md]:
```suggestion
- If you had already used `custom-hiera.yml` and starting to use the tuned profiles, you may want to review the definition of tuned profiles (`/usr/share/foreman-installer/config/foreman.hiera/tuning/common.yaml` and `/usr/share/foreman-installer/config/foreman.hiera/tuning/sizes/`) and remove the duplicated configuration entries from your `custom-hiera.yml`.
```

File: plugins/katello/3.14/installation/index.md
Comment (ekohl) [inline plugins/katello/3.14/installation/index.md]: I'd move all these notes to the bottom of the section, below the table. My reasoning is that docs should go from most relevant to least relevant. Using `custom-hiera.yaml` is unsupported and you are expected to be an advanced user already. That means it should be at the bottom.

File: plugins/katello/3.14/installation/index.md
Comment (ehelms) [reply plugins/katello/3.14/installation/index.md]: I kinda disagree with that idea as this goes back to my whole problem with the naming in the first place. The default tuning profile is common. Common is the default. Common is *always* applied so it is the defaults. When looking at say medium, I should be pointed at the medium profile as thats what makes medium _different_ than the default.

Comment (sthirugn) [reply plugins/katello/3.14/installation/index.md]: I have a design question on the real necessity of `foreman.hiera/tuning/common.yaml` file.
I think we should do one of the two options:
Option 1
**********
1. Strip out the changing parameters of  `foreman.hiera/tuning/common.yaml` and move them to a new `foreman.hiera/tuning/sizes/default.yaml` file.
2. And mention in docs that tuning will be always applied in this order:
   - common.yaml
   - `selected tuning profile`. For e.g., medium.yaml  (Note any configuration here overrides default .yaml)
   - custom-hiera.yaml always precedes all other tuning profiles above.

Option 2
**********
- Rename `foreman.hiera/tuning/common.yaml` to `default.yaml` and move it to `foreman.hiera/tuning/sizes`. Note: No common.yaml anymore.
- And mention in docs that tuning will be always applied in this order:
   - default.yaml
   - `selected tuning profile`. For e.g., medium.yaml  (Note any configuration here overrides default .yaml)
   - custom-hiera.yaml always precedes all other tuning profiles above.

File: plugins/katello/3.14/installation/index.md
Comment (sthirugn) [reply plugins/katello/3.14/installation/index.md]: Good point about not emphasizing too much on `custom-hiera.yaml` for entry-level. Now I created a new sub-section.  Moved the note on custom-hiera.yaml to the end of the first section (not to the very end as per your comment) because the second sub-section is mostly for choosing the right tuned profile.

File: plugins/katello/3.14/installation/index.md
Comment (ekohl) [inline plugins/katello/3.14/installation/index.md]: To me it's clear that this can also be combined with `--tuning <profile>` and I also know that `--noop` implies not saving the answers (including tuning). Is that also clear for the average user?

File: plugins/katello/3.14/installation/index.md
Comment (ekohl) [inline plugins/katello/3.14/installation/index.md:218]: :+1: I like an explicit section here

File: plugins/katello/3.14/installation/index.md
Comment (ekohl) [inline plugins/katello/3.14/installation/index.md]: `the bulk action` looks odd. Did you mean `bulk actions`?
```suggestion
It is difficult to find the exact tuning profile for a specific environment in the first attempt because it depends on various factors like the number of managed hosts, the features used in scale (E.g., Remote Execution), bulk actions on hosts, the total amount of content, amount of host traffic to foreman, etc.  Our recommendation is that you start with the tuning profile guidance as shown in the below table based on the number of managed hosts and scale up your environment as needed.
```

**Review by ekohl — APPROVED** [review]
Looks good to me. Minor questions inline, but those don't need to block merging.

**Comment by sthirugn** [comment]
@ekohl @ehelms thank you for the review comments. I learnt a lot from this PR. I don't have write access to this repo so if any of you can merge this please.

File: plugins/katello/3.14/installation/index.md
Comment (sthirugn) [reply plugins/katello/3.14/installation/index.md]: ah good point, I will update this.

**Comment by sthirugn** [comment]
wait before merging, didnt see the last minor comments from @ekohl

File: plugins/katello/3.14/installation/index.md
Comment (sthirugn) [reply plugins/katello/3.14/installation/index.md]: done

File: plugins/katello/3.14/installation/index.md
Comment (sthirugn) [reply plugins/katello/3.14/installation/index.md]: done

**Comment by sthirugn** [comment]
Updated, ready to merge.

**Review by ekohl — APPROVED** [review]

**Comment by ekohl** [comment]
Thanks!

**Pull request merged**
